aten/src/ATen/core/ivalue_inl.h

```diff
@@ -326,15 +323,25 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   }
 
  private:
-  std::mutex mutex_;
+  void runCbs(std::unique_lock<std::mutex>& lock) {
+    std::vector<std::function<void(void)>> cbs;
+    cbs.swap(callbacks_);
```

Comment:
Two goals:
 - we don't want to hold the lock when running the callback. That can lead to deadlocks (e.g. if the callback calls value() it would deadlock)
 - we need to clear callbacks_ after they're invoked.

Swapping with an empty vector does this. 

The other approach is
  std::vector<std::function<void(void)>> cbs{std::move(callbacks_)};

But the move constructor doesn't explicitly guarantee callbacks_ will be empty after (even though it will be in real life).